ARROW-10135: [Rust] [Parquet] Refactor file module to help adding sources

rdettai opened the pull request:
https://issues.apache.org/jira/browse/ARROW-10135

rdettai commented:
I first moved the footer parsing out of the `SerializedFileReader`. I will now try to move as much logic as possible from the `SerializedFileReader` and `SerializedRowGroupReader` traits to the more generic `FileReader` and `RowGroupReader` traits

rdettai commented:
The discussion with @alamb about the need for an intermediate layer when reading a parquet file is discussed on [JIRA](https://issues.apache.org/jira/browse/ARROW-10135)

The highlights of the current implementation:
- The public API has changed, but keeps working for `File` and `Path` thanks to the corresponding trait implementations. `Cursor` cannot be used any more because it requires data copies when being passed around with `clone()` (this was already the case before in the implem of `TryClone` for `Cursor<Vec<u8>>`).
- I have added a custom cursor type (`SliceableCursor`) that allows to generate cursor slices without cloning the underlying data. This can be used to read in-memory files. I guess it could be made more generic, but this would be for convenience only and I find it simple and clear as is.
- I have separated the implem (`SerializedFileReader`, `SerializedRowGroupReader`...) from the traits (`FileReader`, `RowGroupReader`...) for more clarity. I know that this is not how the code base is structured in the rest of the project but I tend to get lost in these huge files with millions of struct/trait/impl blocks. I'm very much open to suggestion about this point!
- There is nothing about async/parallelism yet, I have to think about it a little bit.

@alamb : can you take a look at the new `ChunckReader` trait and how it is integrated to the rest of the reader? What do you think about it?

alamb commented on rust/parquet/src/file/reader.rs:
```diff
@@ -19,34 +19,48 @@
+
+pub trait ReadChunck: Read + Length {}
+pub trait ReadSeekChunck: ReadChunck + Seek {}
+impl<T: Read + Length> ReadChunck for T {}
```
I like this pattern (of automatically implementing the combined trait for anything that implements `Read` + `Length` (I struggled doing that with FileReader when working with the Parquet API earlier in my Rust days)

alamb commented on rust/parquet/src/util/cursor.rs:
```diff
@@ -0,0 +1,203 @@
+
+/// This is object to use if your file is already in memory.
+/// The sliceable cursor is similar to std::io::Cursor, except that it makes it easy to create "cursor slices".
+/// To achieve this, it uses Rc instead of shared references. Indeed reference fields are painfull
```
I would personally suggest moving to `Arc` here rather than `Rc` across the board in the Parquet reader. I suggest `Arc` to "future proof" the code as `Arc` almost always is needed with `async` and multi-threading (as `Rc` doesn't implement `Send`)

nevi-me replied:
Should we implement this separately? That'd be my preference

sunchao replied:
Yes maybe move this into a separate PR? given the current one is already big enough.

alamb commented on rust/parquet/src/util/io.rs:
```diff
@@ -17,11 +17,31 @@
 
 // ----------------------------------------------------------------------
+
+/// TryClone tries to clone the type and should maintain the `Seek` position of the given
```
👍  for comments

alamb reviewed (commented):
@rdettai  -- I looked at `ChunckReader ` and skimmed how it was used in the rest of the code. 

TLDR is I think it is a nice improvement and in my opinion this PR moves the Rust parquet writer forward. 

Some comments:
1. I think using a more standard spelling of  `Chunk` rather than `Chunck` might improve the readability somewhat
2. I wonder how important separating out the types and`get_read` and `get_read_seek` are? Given the underlying implementation of `ChunckReader` will have to implement seekable streams anyways, I wonder how much more performance is to be gained for the additional complexity in API (indeed both `File` and `SliceableCursor` appear to use the same type for `T` and `U`)
3. I like `SliceableCursor` -- it is a nice way to avoid the copies that occur with the actual Cursor implementation
4. I think we should begin moving the parquet reader towards using `Arc` rather than `Rc` so as to better prepare ourselves for multi-threading

rdettai commented:
Thanks a lot for your time and comments! 

> using a more standard spelling of Chunk rather than Chunck

You mean "a more correct spelling" 😉 

> I wonder how important separating out `get_read` and `get_read_seek` is

Right, I was just looking at this. There might be some simplifications that could be implemented here.

> I think we should begin moving the parquet reader towards using Arc

That is true, but there is a major problem with the way `FileSource` works in that case. The fact that we share a single file handle between instances (because it is the behavior of `File::clone()`) prevents us from accessing concurrently a same file.

Give me a bit of time, I will try to fix these. I'll come back to you if I have further doubts. Thanks again !!!

rdettai commented:
@alamb I think I addressed most of your concerns. The only one that remains is the necessity to prepare for async, but I have digged a little bit into and I think that tackling this properly will require work that is not really in the scope of this PR. There is more to it than converting Rc to Arc because of this shared handle to the file. Lets do that work in an other PR associated to a sub-tasks of [ARROW-9464](https://issues.apache.org/jira/browse/ARROW-9464) or simply associated to [ARROW-9674](https://issues.apache.org/jira/browse/ARROW-9674). I will work on it next week, but until then we can validate and merge this one.

alamb commented:
> The only one that remains is the necessity to prepare for async, but I have digged a little bit into and I think that tackling this properly will require work that is not really in the scope of this PR

I agree

alamb commented on rust/parquet/src/file/reader.rs:
```diff
@@ -18,35 +18,37 @@
+
+use crate::column::reader::ColumnReaderImpl;
+
+/// Length should return the amount of bytes that implementor contains.
```
```suggestion
/// Length should return the total number of bytes in the input source.
```

alamb commented on rust/parquet/src/file/reader.rs:
```diff
@@ -18,35 +18,37 @@
+    fn len(&self) -> u64;
+}
+
+/// The ChunkReader trait generates readers of chunks of a source.
```
👍

alamb commented on rust/parquet/src/file/reader.rs:
```diff
@@ -18,35 +18,37 @@
+    fn get_read(&self, start: u64, length: usize) -> Result<Self::T>;
+}
 
 // ----------------------------------------------------------------------
```
For backwards compatibility in other projects, I think we need to allow  `use parquet::file::reader::TryClone` to keep working (`TryClone` moved to `util::io::TryClone`, and appears not to be public anymore). 

Perhap something like this (untested):
```suggestion
pub use super::TryClone;
// ----------------------------------------------------------------------
```

When I tried to compile an in-house project with this branch I got the following error: 

```

error[E0432]: unresolved import `delorean_parquet::parquet::file::reader::TryClone`
  --> delorean_parquet/src/lib.rs:13:28
   |
13 |     file::reader::{Length, TryClone},
   |                            ^^^^^^^^ no `TryClone` in `parquet::file::reader`

error[E0432]: unresolved import `delorean_parquet::parquet::file::reader::TryClone`
```

When I tried to use the copy in `util::io` I got: 

```
error[E0603]: module `util` is private
  --> delorean_parquet/src/lib.rs:14:5
   |
14 |     util::io::{TryClone},
   |     ^^^^ private module
   |

```

When I used the import in `seriailized_reader` I got a similar error: 

```
error[E0603]: trait `TryClone` is private
  --> delorean_parquet/src/lib.rs:14:30
   |
14 |     file::serialized_reader::TryClone,
   |                              ^^^^^^^^ private trait
   |
```

nevi-me replied:
@alamb is this still an issue?

alamb replied:
I think it is (as the `ParquetWriter` uses `TryClone`) but I may not have tested with the latest changes from @rdettai

alamb commented on rust/parquet/src/util/cursor.rs:
```diff
@@ -0,0 +1,203 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
I discovered above that the entire `util` module is not public -- 

https://github.com/apache/arrow/blob/master/rust/parquet/src/lib.rs#L39

```
#[macro_use]
mod util;
```

So none of these pub traits can be used outside this crate at the moment

rdettai replied:
this is intended

alamb reviewed (commented):
I tried to compile an internal project that uses this branch against an internal project we have (that uses `SerializedFileReader` with something other than `File` and I hit several compiler errors which I haven't fully worked through. 

The first is with `TryClone` being private to the crate (comments inline below) though upon further reading of this PR it seems like the right thing for me to have done is to change the struct we are using to implement ChunkReader instead of TryClone, Length, etc.


Another error is the following:
```
error[E0599]: no method named `metadata` found for struct `parquet::file::serialized_reader::SerializedFileReader<R>` in the current scope
   --> delorean_parquet/src/metadata.rs:46:35
    |
46  |     let parquet_metadata = reader.metadata();
    |                                   ^^^^^^^^ private field, not a method
```

Sadly I ran out of time this morning to keep looking at this. I will try and spend some more time either later today or perhaps tomorrow.

alamb commented:
Sorry I don't have more time to spend on this -- I am still struggling to get my existing project code to compile against this branch and this PR is pretty massive to digest. 

One other thing I found this morning is that the `TryClone` trait is still used in `ParquetWriter` so I do think it needs to be publically exported somehow (or we need to refactor the writer)

I have also been trying to figure out if the `ParquetReader` trait is still needed after this refactoring.

So I guess all in all my conclusion is that the ideas behind this PR are good but:
1. It is hard to review given its size
2. There will likely be a fairly substantial backwards incompatible change for anyone using the ParquetReader in their project

nevi-me commented:
I'll also only be able to review this in the next 2 week(end)s, and I'd like to go through it before it's merged.

I'm also going to reach out to a few active crates that depend on Parquet, so I can solicit their feedback on these changes.
I anticipate a lot of changes + improvements for the 3.0.0 release, so if we can manage that without causing a lot of breakage; I think we should go for it.

alamb commented:
I should probably make clear that anyone using the `SerializedFileReader` with `File` will likely have no issues (as @rdettai  has done a great job of keeping the that level interface consistent. Any project that uses a custom input source (as we did in my internal project) by implementing `ParquetReader` and associated traits may have more to change.

I actually think the changes in this PR will make such custom input sources easier to write in the future, but existing code will have to be rejiggered.

rdettai commented:
I'm open to moving traits such as `ParquetReader` back to `reader.rs` or re-exposing them with `pub use crate::`. But I have the feeling these traits are very specific to the implementation of the `util::io::FileSource`, which is private because it is mainly a slicing/buffering wrapper around `File` (hence the name ;-)). This wrapper makes sense because File is cheap to clone and needs to be buffered, but this will not be generally the case for other implems of `ParquetReader`. What about your usecase @alamb ? Do you think it is worth it have the ChunkReader be implemented for `T:ParquetReader` rather than `File` ? This would be more flexible, maybe a little bit too much...

It is a good observation that `TryClone` is also used in the writer. I didn't want to move things around in the writer because it would have made the PR even more massive and also because I've been told that there was an active development effort going on over there :-). Similar changes might apply.

alamb commented:
@rdettai  --

> This wrapper makes sense because File is cheap to clone and needs to be buffered, but this will not be generally the case for other implems of ParquetReader. 

Yes, actually that was our experience (when we used a buffered implementation, we found that the underlying (large) buffer got copies around a lot. 

I think in general, the cleaner thing (and my preference) would be to simply port my code to use `ChunkReader` (I think it will end up being cleaner and likely more performant). 

> Do you think it is worth it have the ChunkReader be implemented for T:ParquetReader rather than File ?  

The upside is that it would be somewhat more backwards compatible, but unless other reviewers feel strongly I personally think we should just do the breaking change in terms of `ChunkReader` and move on

alamb commented:
I am but one opinion -- I wonder if @nevi-me  or @sunchao  have any thoughts / opinions about where this PR is heading

nevi-me commented:
> I am but one opinion -- I wonder if @nevi-me or @sunchao have any thoughts / opinions about where this PR is heading

I'm principally fine with the changes, and the direction of the PR. I'll highly likely be working more on the non-Arrow side of Parquet, so I expect that I might be making a lot of changes on top of this.

There seems to be some demand for supporting sources and sinks other than files, and as this helps with that; I'm pro getting it merged.

@rdettai are there still more changes that you intend on making, and @alamb are all your queries and concerns addressed? Thanks for the detailed review.

alamb commented:
@nevi-me  -- I think this is good enough to merge; Once it is merged, I'll try and port my internal project asap (which has custom readers and writers) and make a PR to do any touchups needed (like re-exporting TryClone if need be)

sunchao commented:
From the description the changes proposed here makes sense to me as well - but seems the PR itself is messed up and contains many unrelated changes.

rdettai commented:
> @rdettai are there still more changes that you intend on making, and @alamb are all your queries and concerns addressed? Thanks for the detailed review.
@nevi-me Depends on whether we want `ParquetReader` to remain public or not. If not, I think the PR is fine, otherwise, I can bring it back into `parquet::file::reader`.

@sunchao I tried to restrain myself on this PR 😄. Honestly, I had to move quite a lot of things around because this touches a core API and things were very "monolithic". There are two points were I'm getting a little bit out of the main concern:
- I added a `Seek` implem to the `FileSource` that ended up not being useful. I am removing it right now.
- The typo fix in `array_reader`, but I'm sure you can forgive me that one :-)

sunchao commented:
Thanks @rdettai . I'll take a look at this PR today.

sunchao commented on rust/parquet/src/file/footer.rs:
```diff
@@ -0,0 +1,266 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//! Contains file reader API and provides methods to access file metadata, row group
```
I think this needs to be updated?

sunchao commented on rust/parquet/src/file/serialized_reader.rs:
```diff
@@ -0,0 +1,747 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//! Contains file reader implementations and provides methods to access file metadata, row group
```
Needs update as well - specifically for serialized reader

rdettai replied:
I'll try to come up with something more explicit!

sunchao commented on rust/parquet/src/util/io.rs:
```diff
@@ -17,11 +17,25 @@
+
+/// ParquetReader is the interface which needs to be fulfilled to be able to parse a
+/// parquet source.
+pub trait ParquetReader: Read + Seek + Length + TryClone {}
```
(not related to this PR) I think `ParquetReader` is a little confusing - perhaps a better name can be `ParquetSource` or something. And similarly `ParquetSink` for `ParquetWriter`.

sunchao commented on rust/parquet/src/file/footer.rs:
```diff
@@ -0,0 +1,266 @@
+///
+/// The reader first reads DEFAULT_FOOTER_SIZE bytes from the end of the file.
+/// If it is not enough according to the length indicated in the footer, it reads more bytes.
+pub fn parse_metadata<R: ChunkReader>(chunk_reader: &R) -> Result<ParquetMetaData> {
```
+1 - nice to see this separated into another file

sunchao commented on rust/parquet/src/file/reader.rs:
```diff
@@ -18,35 +18,37 @@
+/// The ChunkReader trait generates readers of chunks of a source.
+/// For a file system reader, each chunk might contain a clone of File bounded on a given range.
+/// For an object store reader, each read can be mapped to a range request.
+pub trait ChunkReader: Length {
```
nit: maybe `ChunkRead`?

sunchao commented on rust/parquet/src/file/reader.rs:
```diff
@@ -18,35 +18,37 @@
+/// Length should return the total number of bytes in the input source.
+/// It's mainly used to read the metadata, which is at the end of the source.
+#[allow(clippy::len_without_is_empty)]
+pub trait Length {
```
maybe io.rs is more suitable for this trait?

rdettai commented:
I'm having second thoughts about the `ChunkReader` interface. It has a length, but what we really want is the capability to "read from end" in order to get the parquet footer. So I'm thinking about removing the Length trait and give to `get_read` the capability to read from end, for instance by making `start: u64` an `enum { FromStart(u64), FromEnd }`. This is particularly interesting when reading from a remote store where getting the size is expensive. What do you think ?

alamb commented:
> I'm having second thoughts about the ChunkReader interface.

I personally think the `ChunkReader` interface is good enough as is and this PR has been hanging out for quite a bit. 

By adding more sophistication (FromStart, FromEnd) you may save an S3 or other object store`  metadata request to get the length in theory, but I suspect most applications will already have the length metadata from other sources anyway (e.g.  because they had to list the objects for some other reason) so acquiring the length may not be as expensive as it first appears.

rdettai commented:
I agree that this PR is hanging, but as this is an API change, I guess its better to think things through before moving forward! 😄 This should maybe have been thought through in a design document rather than in the PR... 😅

Its true that most of the time you will have the length around from an other source. But shouldn't we prefer a traits that contains exactly the minimum amount of information we need to make the `FileReader` implem work ?
Proposition:
```
pub enum ChunkMode {
    FromStart(u64),
    FromEnd(u64),
}

pub trait ChunkReader {
    type T: Read + Length;
    fn get_read(&self, start: ChunkMode, length: usize) -> Result<Self::T>;
}
```

I can have the implem ready by the end of the day!

rdettai commented:
Hi! I did some experimentation with this new interface (with `ChunkMode`). It works but is not flawless. You are right @alamb , we should move on with this PR the way it is currently implemented (with `get_read(&self, start: u64, length: usize)`), and we'll discuss the other possibility in a dedicated thread.
@sunchao I have responded to all your comments. If you are okay with my answers, I think we can merge this.

sunchao reviewed (approved):
LGTM from my side. Thanks @rdettai !

alamb commented:
FYI I have filed https://issues.apache.org/jira/browse/ARROW-10390 for the issue with `TryClone` not being public. I'll try and whip up a PR to get it fixed.
